File: smoke_tests/csharp/csrf.cs
// Cross-Site Request Forgery (CSRF) vulnerabilities in C#
using System;
using System.Web.Mvc;

namespace VulnerableApp
{
    public class CsrfController : Controller
    {
        // Test 1: Missing ValidateAntiForgeryToken
        [HttpPost]
        public ActionResult TransferFunds(decimal amount, string toAccount)
        {
            // VULNERABLE: No CSRF token validation
            ProcessTransfer(amount, toAccount);
            return Ok();
        }

        // Test 2: State change via GET request
        [HttpGet]
        public ActionResult DeleteAccount(int id)
        {
            // VULNERABLE: State change should not be GET
            DeleteUser(id);
            return RedirectToAction("Index");
        }

        // Test 3: Missing token on sensitive action
        [HttpPost]
        public ActionResult ChangePassword(string newPassword)
        {
            // VULNERABLE: Password change without CSRF protection
            UpdatePassword(User.Identity.Name, newPassword);
            return Ok();
        }

        // Test 4: Missing token on email change
        [HttpPost]
        public ActionResult ChangeEmail(string newEmail)
        {
            // VULNERABLE: Email change without CSRF protection
            UpdateEmail(User.Identity.Name, newEmail);
            return Ok();
        }

        // Test 5: AJAX without anti-forgery
        [HttpPost]
        public JsonResult UpdateProfile(string name, string bio)
        {
            // VULNERABLE: AJAX endpoint without CSRF token
            UpdateUserProfile(name, bio);
            return Json(new { success = true });
        }

        // Test 6: Token only checked on some methods
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult CreateOrder(Order order)
        {
            return Ok();
        }

        [HttpPost]
        public ActionResult CancelOrder(int orderId)
        {
            // VULNERABLE: Related action without token
            CancelOrderById(orderId);
            return Ok();
        }

        // Test 7: CORS misconfiguration allowing CSRF
        [HttpPost]
        public ActionResult ApiEndpoint()
        {
            // VULNERABLE: If CORS allows any origin with credentials
            Response.Headers.Add("Access-Control-Allow-Origin", "*");
            Response.Headers.Add("Access-Control-Allow-Credentials", "true");
            return Ok();
        }

        // Test 8: Cookie without SameSite
        public ActionResult SetAuthCookie()
        {
            var cookie = new System.Web.HttpCookie("auth", "value")
            {
                // VULNERABLE: Missing SameSite attribute
                HttpOnly = true,
                Secure = true
                // SameSite not set
            };
            Response.Cookies.Add(cookie);
            return Ok();
        }

        // Test 9: Admin action without CSRF
        [HttpPost]
        [Authorize(Roles = "Admin")]
        public ActionResult PromoteUser(int userId)
        {
            // VULNERABLE: Admin action without CSRF protection
            SetUserRole(userId, "Admin");
            return Ok();
        }

        // Test 10: Form with autocomplete for sensitive data
        public ActionResult PaymentForm()
        {
            // VULNERABLE: No CSRF token in form
            return View();  // View would need @Html.AntiForgeryToken()
        }

        // Test 11: Token in query string
        [HttpPost]
        public ActionResult ProcessPayment(string token)
        {
            // VULNERABLE: Token should be in header or form body
            // Query string tokens can leak via Referer
            if (ValidateToken(token))
            {
                return Ok();
            }
            return Unauthorized();
        }

        // Placeholder methods
        private void ProcessTransfer(decimal amount, string to) { }
        private void DeleteUser(int id) { }
        private void UpdatePassword(string user, string pass) { }
        private void UpdateEmail(string user, string email) { }
        private void UpdateUserProfile(string name, string bio) { }
        private void CancelOrderById(int id) { }
        private void SetUserRole(int id, string role) { }
        private bool ValidateToken(string token) => true;
    }

    public class Order
    {
        public int ProductId { get; set; }
        public int Quantity { get; set; }
    }
}

File: smoke_tests/ruby/csrf.rb
# Cross-Site Request Forgery (CSRF) vulnerabilities in Ruby/Rails
class CsrfController < ApplicationController
  # Test 1: Skipping CSRF protection
  skip_before_action :verify_authenticity_token

  def transfer_funds
    amount = params[:amount]
    to_account = params[:to_account]
    # VULNERABLE: No CSRF protection
    current_user.transfer(amount, to_account)
    redirect_to account_path
  end

  # Test 2: State change via GET
  def delete_account
    # VULNERABLE: DELETE via GET request
    User.find(params[:id]).destroy
    redirect_to root_path
  end

  # Test 3: protect_from_forgery except
  protect_from_forgery except: [:change_password]

  def change_password
    # VULNERABLE: Password change without CSRF
    current_user.update(password: params[:new_password])
    redirect_to root_path
  end

  # Test 4: AJAX without CSRF
  protect_from_forgery with: :null_session

  def update_profile
    # VULNERABLE: null_session allows CSRF
    current_user.update(profile_params)
    render json: { success: true }
  end

  # Test 5: API without CSRF
  skip_before_action :verify_authenticity_token, if: :json_request?

  def api_action
    # VULNERABLE: JSON requests without CSRF
    perform_sensitive_action
    render json: { result: 'done' }
  end

  # Test 6: CORS misconfiguration
  def cors_endpoint
    # VULNERABLE: CORS allows any origin with credentials
    response.headers['Access-Control-Allow-Origin'] = '*'
    response.headers['Access-Control-Allow-Credentials'] = 'true'
    render json: { data: 'sensitive' }
  end

  # Test 7: Cookie without SameSite
  def set_cookie
    # VULNERABLE: No SameSite attribute
    cookies[:auth] = {
      value: 'token',
      httponly: true,
      secure: true
      # SameSite not set
    }
  end

  # Test 8: Admin action without CSRF
  skip_before_action :verify_authenticity_token, only: [:promote_user]

  def promote_user
    # VULNERABLE: Admin action without protection
    User.find(params[:id]).update(role: 'admin')
    redirect_to users_path
  end

  # Test 9: Token in query string
  def process_payment
    token = params[:csrf_token]
    # VULNERABLE: Token in URL can leak via Referer
    if valid_token?(token)
      process_payment_internal
    end
  end

  # Test 10: protect_from_forgery disabled conditionally
  protect_from_forgery unless: -> { request.format.json? }

  def sensitive_json_action
    # VULNERABLE: JSON requests bypass CSRF
    perform_action
    render json: { status: 'ok' }
  end

  private

  def json_request?
    request.format.json?
  end

  def profile_params
    params.require(:profile).permit(:name, :bio)
  end

  def valid_token?(token)
    # Simplified validation
    true
  end

  def perform_sensitive_action; end
  def perform_action; end
  def process_payment_internal; end
end
